localstack/testing/pytest/validation_tracking.py

```diff
@@ -0,0 +1,138 @@
+
+Keeping a record of how long ago a test was validated last,
+we can periodically re-validate ALL AWS-targeting tests (and therefore not only just snapshot-using tests).
+"""
+
+import datetime
+import json
+import os
+from typing import Optional
+
+import pluggy
+import pytest
+
+from localstack.testing.aws.util import is_aws_cloud
+
+
+def find_snapshot_for_item(item: pytest.Item) -> Optional[dict]:
+    base_path = os.path.join(item.fspath.dirname, item.fspath.purebasename)
+    snapshot_path = f"{base_path}.snapshot.json"
+
+    if not os.path.exists(snapshot_path):
+        return None
+
+    with open(snapshot_path, "r") as fd:
+        file_content = json.load(fd)
+        return file_content.get(item.nodeid)
+
+
+def find_validation_data_for_item(item: pytest.Item) -> Optional[dict]:
+    base_path = os.path.join(item.fspath.dirname, item.fspath.purebasename)
+    snapshot_path = f"{base_path}.validation.json"
+
+    if not os.path.exists(snapshot_path):
+        return None
+
+    with open(snapshot_path, "r") as fd:
+        file_content = json.load(fd)
+        return file_content.get(item.nodeid)
+
+
+def record_passed_validation(item: pytest.Item, timestamp: Optional[datetime.datetime] = None):
+    base_path = os.path.join(item.fspath.dirname, item.fspath.purebasename)
+    with open(f"{base_path}.validation.json", "w+") as fd:
+        # read existing state from file
+        try:
+            content = json.load(fd)
+        except json.JSONDecodeError:  # expected on first try (empty file)
+            content = {}
+
+        # update for this pytest node
+        if not timestamp:
+            timestamp = datetime.datetime.now(tz=datetime.timezone.utc)
+        content[item.nodeid] = {"last_validated_date": timestamp.isoformat(timespec="seconds")}
+
+        # save updates
+        fd.seek(0)
+        json.dump(content, fd, indent=2)
+
+
+# TODO: we should skip if we're updating snapshots
+# make sure this is *AFTER* snapshot comparison => tryfirst=True
+@pytest.hookimpl(hookwrapper=True, tryfirst=True)
+def pytest_runtest_call(item: pytest.Item):
+    outcome: pluggy.Result = yield
+
+    # we only want to track passed runs against AWS
+    if not is_aws_cloud() or outcome.excinfo:
+        return
+
+    record_passed_validation(item)
+
+
+# this is a sort of utility used for retroactively creating validation files in accordance with existing snapshot files
+# it takes the recorded date from a snapshot and sets it to the last validated date
+# @pytest.hookimpl(trylast=True)
+# def pytest_collection_modifyitems(session, config, items: list[pytest.Item]):
+#     for item in items:
+#         snapshot_entry = find_snapshot_for_item(item)
+#         if not snapshot_entry:
+#             continue
+#
+#         snapshot_update_timestamp = datetime.datetime.strptime(snapshot_entry["recorded-date"], "%d-%m-%Y, %H:%M:%S").astimezone(tz=datetime.timezone.utc)
+#
+#         record_passed_validation(item, snapshot_update_timestamp)
```

Comment:
Just a minor thing about the collection and loading the file for each item, but it feels like a way too early optimization: would there be a way to group the items by `base_path` and load the file only once? I feel like it would over-complicate things quite a bit for maybe not so much gain.
But looking at the S3 snapshot file, it has 218 entries, and loading this 340kb file 218 times feels like a lot. 

Maybe using `itertools.groupby` could help a bit somehow, I've given it a quick try and I could get the time to collect all snapshots from 850ms to 30ms. 

```python

def get_snapshot_file(base_path: str) -> Optional[dict]:
    snapshot_path = f"{base_path}.snapshot.json"

    if not os.path.exists(snapshot_path):
        return None

    with open(snapshot_path, "r") as fd:
        file_content = json.load(fd)
        return file_content


def find_snapshot_for_item(item: pytest.Item, snapshot_data: dict = None) -> Optional[dict]:
    if not snapshot_data:
        base_path = os.path.join(item.fspath.dirname, item.fspath.purebasename)
        snapshot_path = f"{base_path}.snapshot.json"

        if not os.path.exists(snapshot_path):
            return None

        with open(snapshot_path, "r") as fd:
            snapshot_data = json.load(fd)

    return snapshot_data.get(item.nodeid)


@pytest.hookimpl(trylast=True)
def pytest_collection_modifyitems(session, config, items: list[pytest.Item]):
    def sort_key(x: pytest.Item):
        return os.path.join(x.fspath.dirname, x.fspath.purebasename)

    for key, g_items in groupby(sorted(items, key=sort_key), key=sort_key):
        snapshot_data = get_snapshot_file(base_path=key)
        if not snapshot_data:
            continue
        for item in g_items:
            snapshot_entry = find_snapshot_for_item(item, snapshot_data)
            if not snapshot_entry:
                continue
```

It's just a thing I've thought about while reviewing, it's not so much linked to this PR but more to the collection, but the gain is so small that it feels quite overkill for no reason. I just gave it a try, so I thought it was worth sharing anyway 😅

Reply:
You absolutely could, but you also correctly identified that it's basically pointless because of the scale, in which case I prefer keeping this as straight-forward as possible for now. These things are basically only executed on-demand on your local machine for now. The commented out area especially is literally something that we only do once for community and once for pro now :sweat_smile: 

Still a good exercise to think about, so feel free to go ahead and add a follow-up PR in the future. I'd be glad to share some ownership over the validation process :wink: